### ISSUE TITLE
Non-integers should be allowed for exp, nbf, iat

### ISSUE THREAD

**Issue by ionparticle** [web-token/jwt-framework#264]
I had to interact with an app that decided to use a float for their exp. Turns out this is technically valid. RFC7519 specifies exp, nbf, iat as [NumericDates](https://tools.ietf.org/html/rfc7519#page-6), and it defines NumericDates as allowing non-integers:

>This is equivalent to the IEEE Std 1003.1,
> 2013 Edition [POSIX.1] definition "Seconds Since the Epoch", in
> which each day is accounted for by exactly 86400 seconds, other
> than that **non-integer values can be represented**

Suggest combining `is_float()` with `is_int()` or maybe use [is_numeric()](https://www.php.net/manual/en/function.is-numeric.php):
https://github.com/web-token/jwt-framework/blob/953121d5f39a3e8d4ff42646b3a75995d16c4b01/src/Component/Checker/ExpirationTimeChecker.php#L48

**Comment by Spomky** [comment]
Hi @ionparticle,

You are right, non-numeric values can be used for claims `exp`, `nbf` or `iat`.
However to avoid spaghetti code for less common representations of these claims, the provided Checker objects only accept an integer.
Nevertheless, you are free to implement your own object with that specific use case.
When done, you just have to use this custom object instead of the provided one.

```php
<?php

declare(strict_types=1);

namespace Acme\Checker;

use function is_int;
use function is_string;

/**
 * This class is a claim checker.
 * When the "exp" is present, it will compare the value with the current timestamp.
 */
final class CustomExpirationTimeChecker implements ClaimChecker, HeaderChecker
{
    private const NAME = 'exp';

    /**
     * @var int
     */
    private $allowedTimeDrift;
    /**
     * @var bool
     */
    private $protectedHeaderOnly;

    public function __construct(int $allowedTimeDrift = 0, bool $protectedHeaderOnly = false)
    {
        $this->allowedTimeDrift = $allowedTimeDrift;
        $this->protectedHeaderOnly = $protectedHeaderOnly;
    }

    /**
     * {@inheritdoc}
     *
     * @throws InvalidClaimException if the claim "exp" is not valid
     */
    public function checkClaim($value): void
    {
        if (is_int($value)) {
            if (time() > $value + $this->allowedTimeDrift) {
                throw new InvalidClaimException('The token expired.', self::NAME, $value);
            }
            return;
        }
        if (is_string($value) {
           //Your custom implementation goes here
        }
        throw new InvalidClaimException('"exp" must be an integer.', self::NAME, $value);
    }

    public function supportedClaim(): string
    {
        return self::NAME;
    }

    /**
     * @param mixed $value
     *
     * @throws InvalidHeaderException if the claim "exp" is not valid
     */
    public function checkHeader($value): void
    {
        if (is_int($value)) {
            if (time() > $value + $this->allowedTimeDrift) {
                throw new InvalidHeaderException('The token expired.', self::NAME, $value);
            }
            return;
        }
        if (is_string($value) {
           //Your custom implementation goes here
        }
        throw new InvalidHeaderException('"exp" must be an integer.', self::NAME, $value);
    }

    public function supportedHeader(): string
    {
        return self::NAME;
    }

    public function protectedHeaderOnly(): bool
    {
        return $this->protectedHeaderOnly;
    }
}
```

**Comment by ionparticle** [comment]
Oh no, my issue isn't with non-numeric values (i.e. strings). The way I read the spec, strings aren't allowed anyways. The issue is with floating point values, e.g.: {"exp": 1596044815.83}

**Comment by Spomky** [comment]
`is_int`/`is_float` are now used for these objects.
Available in `v2.2.4`
